### DISCUSSION TITLE
Dropping "remote content" design pattern

### DISCUSSION THREAD

**Discussion by subpop** [RedHatInsights/yggdrasil#213] [Ideas]
Worker authors are seeking additional ways to send data back from a worker to hosted services. The design of the data routing methods currently assumes a single pair of transmit/receive methods: `com.redhat.Yggdrasil1.Dispatcher1.Transmit` sends data from a worker to hosted services via the configured transport if a worker's `com.redhat.Yggdrasil1.Worker1.RemoteContent` property is false or via an HTTP POST if the property is true. Similarly, when yggd receives a data message on its "data/in" topic, it will route the data if the property is false or will perform an HTTP GET to a URL if the property is true.

These two pathways are mutually exclusive, and are made at development time by the worker authors. There is no method to switch the value at runtime. There is no means for a worker to use *both* at appropriate times.

I propose we do the following:
* Drop the `com.redhat.Yggdrasil1.Worker1.RemoteContent` property
* Add a new method to `com.redhat.Yggdrasil1.Dispatcher1` called `Post` with a signature as follows: `func Post(addr string, headers map[string]string, data []byte) (code int, headers map[string]string, body []byte)`. This method will send an authenticated HTTP POST request to the URL specified by `addr`, setting the values in `headers` to the request headers and `body` to the request body. The HTTP response is then returned in the 3 return values.
* Add a new method to `com.redhat.Yggdrasil1.Dispatcher` called `Get` with a signature as follows: `func Get(addr string, headers map[string]string) (code int, headers map[string]string, body []byte)`. This method will send an authenticated HTTP GET request to the URL specified by `addr`, setting the values in `headers` to the request headers. The HTTP response is then returned in the 3 return values.
* Refactor the "data/in" topic handler dropping support for "RemoteContent".
* Refactor `com.redhat.Yggdrasil1.Dispatcher1.Transmit`, dropping support for "RemoteContent".

These new methods allow workers to decide when and how to return data to hosted services. If the worker wants to send a small message, it can invoke `com.redhat.Yggdrasil1.Dispatcher1.Transmit`; the data will be publishes across MQTT. If the worker wants to send a large amount of data, it can invoke `com.redhat.Yggdrasil1.Dispatcher1.Post`; the data will be sent in an HTTP request to a URL.

This does change the way a worker formerly configured to use `RemoteContent` will get its data. Under the current design, the worker is handed data in a single D-Bus method invocation of the `com.redhat.Yggdrasil1.Worker1.Dispatch` method. Under this proposal, a worker will receive a URL as the message content of its `com.redhat.Yggdrasil1.Worker1.Dispatch` method. It then has to make a second D-Bus method call, invoking `com.redhat.Yggdrasil1.Dispatcher1.Get` to retrieve the contents of the URL.

**Comment by jirihnidek** [comment]
Hi @subpop,
I cannot see any property `com.redhat.Yggdrasil1.Worker1.DetachedContent`. I also cannot see anything like `DetachedContent` in source code of yggdrasil. Do you mean `com.redhat.Yggdrasil1.Worker1.RemoteContent`?

Is it necessary to introduce two new methods `Get` and `Post`. Wouldn't be better to introduce only one generic method `Request`? It would allow to call other methods like `PUT`, `DELETE`, `HEAD`, etc. Or do you have any reason why you want to restrict methods only to `GET` and `POST`?

**Reply by subpop** [reply]
> Hi @subpop, I cannot see any property `com.redhat.Yggdrasil1.Worker1.DetachedContent`. I also cannot see anything like `DetachedContent` in source code of yggdrasil. Do you mean `com.redhat.Yggdrasil1.Worker1.RemoteContent`?

Yes, sorry. I forgot that I renamed it. I'll edit the original post to use the intended property name.

> Is it necessary to introduce two new methods `Get` and `Post`. Wouldn't be better to introduce only one generic method `Request`? It would allow to call other methods like `PUT`, `DELETE`, `HEAD`, etc. Or do you have any reason why you want to restrict methods only to `GET` and `POST`?

I thought about this. It would seem like a reasonable next step, but I was not sure if we *want* to allow workers to send any HTTP verb they want. Yesterday, I felt like it would make more sense to restrict them to only GET and POST. Today, I think a general-purpose `Request` or `Do` method makes more sense. We could always fail the HTTP request if a verb other than `GET` or `POST` is requested.

**Comment by subpop** [comment]
I implemented a draft of this proposal as #223.
